### ISSUE TITLE
Add strategy to track NFTs curated on present materials

### ISSUE THREAD

**Issue by neatonk** [neume-network/strategies#256]
My comments adapted from [discord](https://discord.com/channels/844616619979440168/958806793956524122/1014995234033238087):
- The right way to do this for neume would likely involve using the `call-block-logs` strategy. Since the zora drops contracts are deployed by a factory, they would be handled very much like `sound` is currently.
- This is still WIP and has run into some issues... See https://github.com/neume-network/strategies/issues/252
- Another way to do this would be to use the subgraph provided by Zora: https://thegraph.com/hosted-service/subgraph/iainnash/zora-editions-mainnet
Since much effort has gone into removing the web3subgraph I imagine, using `call-block-logs` would be preferred.
-In either case, the result would be a new platform (zora-drops?) including every drops contract deployed by the Zora drops factory.
- The next hiccup would be in defining present materials as a platform. Since all of the NFTs are curated, they will likely be associated with another platform already.

**Comment by neatonk** [comment]
Yet another way to do this would be to get the contract addresses from the present materials contract first and then get metadata for each contract. This approach could lead to a situation where the same contract is reached by multiple strategies—leading to multiple outputs with potentially differing results.

**Comment by TimDaub** [comment]
> * Since much effort has gone into removing the `web3subgraph` I imagine, using `call-block-logs` would be preferred.

yes. But we're open to use other data sources too.

**Comment by TimDaub** [comment]
- song camps nft contracts https://github.com/ourzora/nft-editions

**Comment by neatonk** [comment]
> - song camps nft contracts https://github.com/ourzora/nft-editions 

Yep. To be clear this only applies to the contracts curated on https://www.presentmaterial.xyz/ and not other songcamp projects.

**Comment by neatonk** [comment]
Implementation notes:
- On the deployed `SingleEditionMintableCreator` contract...
  - listen for `CreatedEdition` event to get the new `editionId` (not an address).
  - call `getEditionAtId` with the new `editionId` to get the contract address.

**Comment by neatonk** [comment]
Steps to add zora editions strategies:
- Add `zora-editions-filter-contracts` strategy to https://github.com/neume-network/strategies. This should be similar to `soundxyz-filter-contracts` strategy, which has a transformer, but no extractor. Follow the [implementation notes](https://github.com/neume-network/strategies/issues/256#issuecomment-1239922568) above to filter logs and get the contract address for each edition. **Note**: This could be split into two strategies: one which filters the logs and one which calls `getEditionAtId` to get the contract addresses.
- Add simple `zora-editions-call-tokenuri` and `zora-editions-get-tokenuri` strategies using the provided [strategy-factories](https://github.com/neume-network/strategies/tree/main/src/strategy-factories).

Steps to include the zora editions strategies in the active crawl path.
- Add `zora-editions-call-tokenuri` and `zora-editions-get-tokenuri` strategies to crawl path in https://github.com/neume-network/data/blob/main/steps/3.mjs.
- Add a step run `zora-editions-filter-contracts` strategy in https://github.com/neume-network/data/blob/main/scripts/generateBlockLogsCrawlPath. **side note**: this is a bash script, that generates a javascript module. The script is called in https://github.com/neume-network/data/blob/main/.github/workflows/node.js.yml.
- Include `zora-editions-filter-contracts-transformation` results in https://github.com/neume-network/data/blob/main/steps/1.mjs
- Add lines to "Update neume-network/data files" step in `https://github.com/neume-network/data/blob/main/.github/workflows/node.js.yml` to copy relevant files.

**Comment by neatonk** [comment]
@TimDaub Does [this](https://github.com/neume-network/strategies/issues/256#issuecomment-1241267105) look legit to you? I'd like to get started on this tonight or tomorrow morning.

**Comment by TimDaub** [comment]
> Add zora-editions-filter-contracts strategy to https://github.com/neume-network/strategies. This should be similar to soundxyz-filter-contracts strategy, which has a transformer, but no extractor. Follow the https://github.com/neume-network/strategies/issues/256#issuecomment-1239922568 above to filter logs and get the contract address for each edition.

I've looked into this and I think it sounds legit. https://etherscan.io/address/0x91A8713155758d410DFAc33a63E193AE3E89F909 This contract essentially emits the edition creation event and basically a different contract handles the edition to tokenId translation, right?
As you described to implement it would be a good first step. Happy to advise whereever I can!

**Comment by neatonk** [comment]
> I've looked into this and I think it sounds legit. 

Thanks

> This contract essentially emits the edition creation event and basically a different contract handles the edition to tokenId translation, right?

For whatever reason, the `getEditionAtId` funciton is on the proxy contract instead of the implementation, so you won't see it on etherscan unless you look at the "Read Contract" tab...
![Screen Shot 2022-09-09 at 3 13 59 PM](https://user-images.githubusercontent.com/21178/189426865-5bd0cef1-f74b-4346-8feb-56cba328a613.png)

**Comment by neatonk** [comment]
Getting started now on the steps described here: https://github.com/neume-network/strategies/issues/256#issuecomment-1241267105

**Comment by TimDaub** [comment]
> Getting started now on the steps described here: [#256 (comment)](https://github.com/neume-network/strategies/issues/256#issuecomment-1241267105)

awesome, looking forwars

**Comment by neatonk** [comment]
@TimDaub Are non-audio NFTs filtered globally or per strategy? If not filtered, most NFTs produced by this strategy will be images.

**Comment by TimDaub** [comment]
sound, catalog v2 and mintsongs only have music NFTs. And for zora, we filter by catalog v1 ids

**Comment by neatonk** [comment]
The details provided above were based on the wrong set of contracts. The following values should be used insetad.
- The contract contracts source code is here: https://github.com/ourzora/zora-drops-contracts 
- The contract creation events are emitted by: https://etherscan.io/address/0xf74b146ce44cc162b601dec3be331784db111dc1#events
- The event topic to filter by is: 0xad59ebba8bfb06ba01a615a611467ca3bef86a275bd5e9704d3b295112550ba5
- The `editionContractAddress` is the next topic in the event.

**Comment by neatonk** [comment]
The `tokenURI` method on the drops contracts returns a data URI. To address this I'll:
- add a 'data-uri' message type to the extraction worker message schema
- add a handler for the data-uri type using https://www.npmjs.com/package/data-uri-to-buffer

**Comment by TimDaub** [comment]
> The `tokenURI` method on the drops contracts returns a data URI.

like this, right:

```
data:application/json;base64,eyJuYW1lIjogIlByZXZpYSB4IENyYXRlciBMYWtlIDEiLCAiZGVzY3JpcHRpb24iOiAiVG95b3RhIFByZXZpYSBhdCBDcmF0ZXIgTGFrZSBOYXRpb25hbCBQYXJrXG5BdWd1c3QgMzEsIDIwMjAiLCAiaW1hZ2UiOiAiaXBmczovL2JhZnliZWliYzM2dTc0aW00ZDJ4Z3licXp6N2Y3bHZ1Z2NrZXg3eWpwaTN5bTNqbGhuZXF2M2VjbDZxP2lkPTEiLCAicHJvcGVydGllcyI6IHsibnVtYmVyIjogMSwgIm5hbWUiOiAiUHJldmlhIHggQ3JhdGVyIExha2UifX0=
```

**Comment by neatonk** [comment]
Yes, although I see that the [ZoraNFTCreatorV1](https://github.com/ourzora/zora-drops-contracts/blob/main/src/ZoraNFTCreatorV1.sol) contract has `createDrop` and `createEdition` methods which use different metadata renderers. One, which returns a data-uri as described above and one which uses a base uri and token id. The base uri can potentially use any scheme.

This can be addressed in the strategy by looking the tokenURI scheme to determine the message type. Or, it can be addressed in the extraction worker by adding a message type that can handle any URI. I'm leaning towards the latter as it a more generic solution.

**Comment by neatonk** [comment]
The`zora-drops-get-tokenuri` strategy requires some thought. As mentioned above, the uri shceme and metadata format may vary. 

For now, I think it is good enough to handle drops which use the [EditionMetadataRenderer](https://github.com/ourzora/zora-drops-contracts/blob/main/src/metadata/EditionMetadataRenderer.sol) and ignore the rest. This is helpful because the uri scheme and metadata format are known and can be transformed to match the required schema. Also, this covers all editions created using the https://create.zora.co/create/edition interface and is sufficient to handle the present materials curated contracts.
